src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,331 @@
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="addr">Account</param>
+        /// <param name="till">specified height</param>
+        /// <returns>Whether deposit lock height was successfully updated.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
+            if (till < Ledger.CurrentIndex(engine.SnapshotCache) + 2) return false; // deposit must be valid at least until the next block after persisting block.
+            Deposit deposit = GetDepositFor(engine.SnapshotCache, addr);
+            if (deposit is null) return false;
+            if (till < deposit.Till) return false;
+            deposit.Till = till;
+
+            PutDepositFor(engine, addr, deposit);
+            return true;
+        }
+
+        /// <summary>
+        /// ExpirationOf returns deposit lock height for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit lock height of the specified address.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public uint ExpirationOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Till;
+        }
+
+        /// <summary>
+        /// BalanceOf returns deposited GAS amount for specified address.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        /// <returns>Deposit balance of the specified account.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public BigInteger BalanceOf(DataCache snapshot, UInt160 acc)
+        {
+            Deposit deposit = GetDepositFor(snapshot, acc);
+            if (deposit is null) return 0;
+            return deposit.Amount;
+        }
+
+        /// <summary>
+        /// Withdraw sends all deposited GAS for "from" address to "to" address. If "to"
+        /// address is not specified, then "from" will be used as a sender.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">From Account</param>
+        /// <param name="to">To Account</param>
+        /// <returns>Whether withdrawal was successfull.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private async ContractTask<bool> Withdraw(ApplicationEngine engine, UInt160 from, UInt160 to)
+        {
+            if (!engine.CheckWitnessInternal(from)) throw new InvalidOperationException(string.Format("Failed to check witness for {0}", from.ToString()));
```

Comment:
@AnnaShaleva A notary node can drain the contract, should be checked that from is != to this.Hash

Reply:
`verify` checks for `WitnessScope.None`.

Reply:
Notary contract is allowed to be present in the signer's list only with `None` scope:
https://github.com/neo-project/neo/blob/bb153eb993d963b1a19336a1bfd26ed877053b9d/src/Neo/SmartContract/Native/Notary.cs#L108